File: sol/solution/0001-0100/0032. Longest Valid Parentheses/Solution.c
int longestValidParentheses(char *s) {
    int n = (int)strlen(s), ans = 0, left = 0, right = 0;
    for (int i = 0; i < n; ++i) {
        if (s[i] == '(') left++; else right++;
        if (left == right) { if (2*right > ans) ans = 2*right; }
        else if (right > left) left = right = 0;
    }
    left = right = 0;
    for (int i = n - 1; i >= 0; --i) {
        if (s[i] == ')') right++; else left++;
        if (left == right) { if (2*left > ans) ans = 2*left; }
        else if (left > right) left = right = 0;
    }
    return ans;
}

File: sol/solution/0001-0100/0032. Longest Valid Parentheses/Solution.rs
impl Solution {
    pub fn longest_valid_parentheses(s: String) -> i32 {
        let bytes = s.as_bytes();
        let mut ans = 0i32;
        let (mut left, mut right) = (0, 0);
        for &b in bytes {
            if b == b'(' { left += 1 } else { right += 1 }
            if left == right { ans = ans.max(2 * right as i32); }
            else if right > left { left = 0; right = 0; }
        }
        left = 0; right = 0;
        for &b in bytes.iter().rev() {
            if b == b')' { right += 1 } else { left += 1 }
            if left == right { ans = ans.max(2 * left as i32); }
            else if left > right { left = 0; right = 0; }
        }
        ans
    }
}